### PR TITLE
Add smart card slot locking functionality

### PR DESCRIPTION
**Pull request by comps**
As the firmware comment says,
```
// lockCard()
//
// Locks the specified card slot, making another lock operation on the same slot
// fail if attempted. An unlock must be issued on the slot to restore successful
// locking functionality.
//
// The idea is to limit accidental use in unexpected scenarios, bricking cards
// by a client entering a wrong password repeatedly due to an automation bug.
// The client can lock a slot before using it and unlock it only if everything
// went smoothly. Any infrastructure crash or automation bug should leave it
// locked, letting a human investigate the situation and manually reset the card
// password attempts using an external card reader and a correct password.
```
this is mainly to prevent bricking cards with a limited wrong-passphrase retry count.

The EEPROM wear leveling code is not perfect, but good enough for basically unlimited endurance. Atmel AVR MCUs can handle ~100k guaranteed writes on each byte, and with tests running 24/7, 100 runs a day, it could get there in ~3 years. Wear-leveling it increases that by 512/2=256 times, or ~768 years.  
Not every ATmega has 1024 bytes of EEPROM, so I wanted to leave this at 512 instead of the 32U4's (Micro) 1024 bytes. Honestly, it doesn't matter in practice.

Dealing with "indexed commands" was done via macros as the rest of the code seemed to use functions for "bigger things", I didn't see utility "inlined" one-liners in the file. To keep it small, self-contained and fast (ability to use compile-time `sizeof()` rather than `strlen()`), and also given that `command` is a global variable, I used macros.  
In general, I have tried to make the new code match the old where possible, sometimes slightly altering the old code as well, but overall keeping the changeset as small as possible.

The best way to review this PR is probably going commit-by-commit, as that helps constrain the scope of each change in one's mind. Also, some commits have extra description / explanation.

---

I tested most of the functionality via the serial line directly (`screen /dev/ttyACM0 115200`), but I did try to make the python CLI robust as well.
```
$ rminator 
{'current': 0, 'present': [1, 2, 4], 'locked': []}
$ rminator insert_card 1
Switching slot to 1
$ rminator 
{'current': 1, 'present': [1, 2, 4], 'locked': []}
$ rminator insert_card 2
Switching slot to 2
$ rminator 
{'current': 2, 'present': [1, 2, 4], 'locked': []}
$ rminator lock_card 1
Locking card slot 1
$ rminator 
{'current': 2, 'present': [1, 2, 4], 'locked': [1]}
$ rminator insert_card 1
Switching slot to 1
...
removinator.removinator.LockedError: Slot 1 is locked
$ rminator lock_card 1
Locking card slot 1
...
removinator.removinator.LockedError: Slot 1 is already locked
$ rminator unlock_card 1
Unlocking card slot 1
$ rminator 
{'current': 2, 'present': [1, 2, 4], 'locked': []}
$ rminator insert_card 1
Switching slot to 1
$ rminator 
{'current': 1, 'present': [1, 2, 4], 'locked': []}
```

### CODE REVIEW

**Comment by comps** [comment]
Known issues (that I'll address together with any other comments):
* `valid_commands` in the python code is unused, it was for an earlier version using `.add_argument(choices=[...])`

**Review by nkinder — APPROVED** [review]

File: client/removinator/removinator.py [hunk 1]
```diff
@@ -159,6 +164,48 @@ def remove_card(self):
...
+        :type slot: int
+        :raises: :exc:`SlotError`, :exc:`CommandError`, :exc:`ValueError`
+        """
+
+        if slot < 0 or slot > 8:
+            raise ValueError('Slot must be between 1 and 8')
+
+        self.send_command('UNLOCK{0}'.format(slot))
+
     def get_status(self):
```
Comment (nkinder) [inline client/removinator/removinator.py:209]: It would be good to also  update the docstring for get_status() to mention that a "locked" key will now be returned  to indicate which card slots are locked.

Comment (nkinder) [reply client/removinator/removinator.py:209]: In general, your changes look really good!  It would be nice to update the "Firmware" section of the main README to briefly describe the new lock/unlock commands.

**Comment by comps** [comment]
Sorry for not fixing it earlier, life (health) got in the way + PTO + busy period after everyone returned from their PTO.

I'll hopefully submit a follow-up PR during the week.

File: client/removinator/removinator.py (hunk 1 above)
Comment (comps) [reply client/removinator/removinator.py:209]: I will probably reword it to point to the source code. This project already suffers from many copies of the same information (in firmware, in python code, in docstrings, in Doxygen, in documentation, etc.).
